stream-parent/stream-client/src/main/java/org/interledger/stream/sender/SimpleStreamSender.java

```diff
@@ -192,70 +293,114 @@ default int totalPackets() {
+        sendMoneyPacketized();
+        return null;
+      });
 
-        // Integer.MAX_VALUE is equal to: 2<sup>31</sup>-1, so break only after we exceed Integer.MAX_VALUE
-        if (sequence.get().compareTo(StreamPacket.MAX_FRAMES_PER_CONNECTION) >= 0) {
-          // Break out of the loop and close the connection. Per IL-RFC-29, "Implementations MUST close the connection
-          // once either endpoint has sent 2^31 packets. According to NIST, it is unsafe to use AES-GCM for more than
-          // 2^32 packets using the same encryption key. (STREAM uses the limit of 2^31 because both endpoints encrypt
-          // packets with the same key.)
-          break;
-        }
+      // To track the duration below...
+      final AtomicReference<Duration> sendMoneyDuration = new AtomicReference<>();
+
+      // All futures will run here using the Cached Executor service.
+      return paymentFuture
+          .whenComplete(($, error) -> {
+            if (error != null) {
+              logger.error("SendMoney Stream failed: " + error.getMessage(), error);
+            }
+            if (!originalAmountToSend.get().equals(this.deliveredAmount.get())) {
+              logger.error("Failed to send full amount");
+            }
+            sendMoneyDuration.set(Duration.between(start, Instant.now()));
+          })
+          .thenApply($ -> {
+            try {
+              return closeConnection(streamConnection.nextSequence());
+            } catch (StreamConnectionClosedException e) {
+              // Return a default close result, even though we didn't _actually_ get to close the STREAM.
+              final ConnectionStatistics connectionCloseResult = ConnectionStatistics.builder()
+                  .amountDelivered(this.deliveredAmount.get())
+                  .numFulfilledPackets(this.numFulfilledPackets.get())
+                  .numRejectPackets(this.numRejectedPackets.get())
+                  .build();
+              logger.error(
+                  "Unable to close Connection that is already closed. connectionCloseResult={} error={}",
+                  connectionCloseResult, e
+              );
+              return connectionCloseResult;
+            }
+          })
+          .thenApply(connectionStatistics -> SendMoneyResult.builder()
+              .amountDelivered(connectionStatistics.amountDelivered())
+              .originalAmount(originalAmountToSend.get())
+              .numFulfilledPackets(connectionStatistics.numFulfilledPackets())
+              .numRejectPackets(connectionStatistics.numRejectPackets())
+              .sendMoneyDuration(sendMoneyDuration.get())
+              .build()
+          );
+    }
+
+    private void sendMoneyPacketized() {
+      final AtomicBoolean timeoutReached = new AtomicBoolean(false);
 
+      timeout.ifPresent($ -> {
+        ScheduledExecutorService timeoutMonitor = Executors.newSingleThreadScheduledExecutor();
```

Comment:
I wonder if this causes a memory leak creating a ScheduledExecutorService but never shutting it down.

Reply:
Confirmed this will actually lead to a memory leak. To avoid this, we should change the scheduled lambda from `timeoutReached.set(true)` to  `{ timeoutReached.set(true); timeoutMonitor.shutdown(); }`